File: 2/reader/MatrixDiagonalReader.cpp
#include <fstream>
#include "MatrixDiagonalReader.h"
#include "../matrix/MatrixDiagonal.h"

Matrix *MatrixDiagonalReader::read(std::ifstream &is) {
    int matrixSize;
    is >> matrixSize;
    MatrixDiagonal *matrix = new MatrixDiagonal(matrixSize);
    for (int i = 0; i < matrixSize; i++) {
        double value;
        is >> value;
        matrix->put(i, i, value);
    }
    return matrix;
}

File: 3/reader/MatrixDiagonalReader.py
from reader.MatrixReader import MatrixReader
import matrix


class MatrixDiagonalReader(MatrixReader):

    def read(self, input_stream):
        matrix_size = int(input_stream.readline())
        new_matrix = matrix.MatrixDiagonal(matrix_size)
        values = list(map(float, input_stream.readline().split()))
        for i in range(matrix_size):
            new_matrix.put(i, i, values[i])
        return new_matrix
